### ISSUE TITLE
Questions about encrypted inference details in tutorial 4

### ISSUE THREAD

**Issue by amyzx** [facebookresearch/CrypTen#283]
Hi, I have two questions about your tutorial 4.

Firstly, I hope to finger out the details for the prediction part.   

This is my understanding. Someone from the CrypTen team can correct me if I am wrong. 
① Alice and Bob firstly encode their model/data. Then they compute different shares **share_0** and **share_1** by `PRZS`. They add shares to their encoded tensor and get the corresponding encrypted model/data. Alice now holds **model_0** and **data_0**, where **model_0 = encoded_model + share_0**.

② Alice does inference and generate **output_0** based on **model_0** and **data_0**. Bob also does similar things. When doing inference, they also need `PRZS` to create some shares. Their output_share are different.

③ Finally, they both call `get_plain_text`, which add all output_shares, and then decode the sum to get the same plaintext float prediction results. 

Secondly, I hope to use CrypTen to do inference on one party (e.g Alice). Assume Alice has already saved all parameters, is it possible for Alice to do inference without interacting with Bob? 

Here, I add some codes which saves encrypted model and data in your tutorial 4. Then I load the saved shares and try to do inference on Alice without interacting with Bob. 
However, I get wrong results with low accuracy. Could you give me some suggestions? Thanks a lot!

    @mpc.run_multiprocess(world_size=2)

    def encrypt_model_and_data():

      rank = comm.get().get_rank() 
  
      # Load pre-trained model to Alice
      plaintext_model = crypten.load_from_party(src_model_path, src=ALICE) 
      # Encrypt model from Alice 
      dummy_input = torch.empty((1, 784))
      private_model = crypten.nn.from_pytorch(plaintext_model, dummy_input)
  
      private_model.encrypt(src=ALICE)  
  
      ######### Save encrypted shares 
      crypten.save(private_model, 'data/model_'+str(rank)+'.pth') 
      crypten.print(f"\nRank {rank}: \n {private_model.state_dict()} \n", in_order = True)
        
      # Load data to Bob
      data_enc = crypten.load_from_party('data/bob_test.pth', src=BOB)
      data_enc2 = data_enc[:count]
      data_flatten = data_enc2.flatten(start_dim=1)
   
      ######### Save encrypted data
      crypten.save(data_flatten, 'data/data_'+str(rank)+'.pth')
      crypten.print(f"\nBob data rank: {rank}: \n {data_flatten._tensor.share[0][:10]} \n", in_order = True)
  
      # Classify the encrypted data
      private_model.eval()
      output_enc = private_model(data_flatten)
  
      ######### Show encrypted output shares
      crypten.print(f"enc output share rank {rank}: \n {output_enc._tensor.share[0]} \n", in_order = True)
        
      # Decrypting the result
      output = output_enc.get_plain_text()
      accuracy = compute_accuracy(output, labels[:count])
      crypten.print("\tAccuracy: {0:.4f}".format(accuracy.item()))
  
      # Obtaining the labels
      pred = output.argmax(dim=1)
      crypten.print("Decrypted labels:\n", pred)
 
    encrypt_model_and_data()
    
    
    print("\nTest do inference in party 0 ")
    X_test = crypten.load("data/data_0.pth") 
    print("X_test")
    # only print the first 10 values to verify the correctness of share
    print(X_test._tensor.share[0][:10])
    
    model = AliceNet() 
    model = crypten.load("data/model_0.pth")
    model.eval() 
    
    Y_test_enc = model(X_test) 
    print("Y_test_enc")
    print(Y_test_enc._tensor.share[0])
    
    # Decrypting the result
    output = Y_test_enc.get_plain_text()
    accuracy = compute_accuracy(output, labels[:count])
    crypten.print("\tAccuracy: {0:.4f}".format(accuracy.item()))
    
    # Obtaining the labels
    pred = output.argmax(dim=1)
    crypten.print("Decrypted labels:\n", pred)

**Comment by knottb** [comment]
This is close to the correct interpretation, but there are some details that create your issue.

1. The encrypted model holds encrypted parameters which are held between ALICE and BOB (each parameter has a share_a and share_b, held by ALICE and BOB respectively).

2. During inference, ALICE and BOB must communicate throughout to compute multiplications. These communications do not reveal any information, but are required in order to perform the desired computations.

3. When you use `crypten.save()` in the code above, you are saving the shares of the data / model parameters. Each of these models is encrypted and cannot be used without having both shares. This is why, when you load `model_0` later, the result is nonsense - the parameters are still encrypted. You may be able to recover a model that works on one party if you call `model.decrypt()` first - but this may not work since CrypTen models must be encrypted. It is possible that you can copy the `state_dict` from you decrypted model to a PyTorch model - but you would need a PyTorch version of AliceNet(). I have known that this cause issues, but haven't had time to check or add support for model decryption.

4. Anything outside of the `@mpc.run_multiprocess(2)` decorated function is run on a single process (i.e. single party). This means we can only run plaintext models / computations here. One should also be careful about how data is used here - in this case there is no data from the mpc function being called in the outer scope since the data is saved / loaded to/from a file. This could be an issue if the data is used directly rather than using a file - but this isn't a problem in the example.

Let me know if this was clear and if you have any other questions.

**Comment by amyzx** [comment]
@knottb  Thanks for your detailed reply! 
I have read some source codes and almost know the details for inference. During inference procedure, all parties still need to communicate many rounds (e.g Beaver protocol) to do multiplication or logical operations on their shares. 

In my code, the party Alice only holds shares of data and model. So when calling forward function, Alice misses the intermidiate data and cannot get the correct result. I still have the following questions.  
1. When reading your paper for CrypTen, you wrote "The Beaver triples are currently provided by the TTP". But I find in tutorial 4, it uses TFP. Are these two providers all availabe?
2. I add `print(tensor.share)` in `PRZS` function, and find sometimes it print empty tensors, which makes me confused. 
3. If I save all intermidiate data which are generated by Beaver protocol or PRZS, and reload the data when needed. Is it possible for Alice to do inference without communicating with Bob? That is, Alice computes the correct result of `Y_test_enc = model(X_test)`.

**Comment by knottb** [comment]
1. Yes, the providers are available. The default is currently TFP (Trusted First Party), but TTP is also available by changing the provider setting. In the current version of the code, you can set the provider using: `crypten.mpc.set_default_provider("TTP")`. Note that this will be changed shortly in version 0.4.0 to have providers set in a `config.yaml` file.

2. As far as I'm aware, `clone()` and `shallow_copy()` should not return empty tensors. We would appreciate if you could provide a small repro for that behavior so we can ensure CrypTen is working properly.

3. Unfortunately ALICE cannot compute inference without BOB unless the model is decrypted. Currently we don't have an easy method for decrypting models in CrypTen, though this is a feature we will be interested in making easier in the future. This is what I was referring to in (3) of my answer above.

**Comment by amyzx** [comment]
@knottb Thanks for your quick reply!

In tutorial 4, the encrypted model has several modules.
![image](https://user-images.githubusercontent.com/8285571/125716437-a4ee2671-fd54-4b2e-80e0-1bd1befdbe12.png)

As for the empty share tensors during inference, I get the following debug information. 
1. During the forward for l1.weight and l2.weight, it does not call PRZS.
2. When transpose encrypted weights in `module 3`, it calls `gradients.py/AutogradPermute`
![image](https://user-images.githubusercontent.com/8285571/125716666-04482994-3341-4e27-9f25-357aeee29b73.png)
3. Then it calls `mpc/mpc.py/_add_regular_function`, and also calls `shallow_copy`.
![image](https://user-images.githubusercontent.com/8285571/125716773-dfea4847-9e54-47b6-8557-47dfc1770727.png)
![image](https://user-images.githubusercontent.com/8285571/125716779-f4ad549a-d9d5-447b-a354-a6c61d645ac2.png)
4. When generating an ampty `MPCTensor`, it calls `__init__`. **Here**, I find that the flag is False. So it will call `mpc/primitives/Arithmetic.py/__init__` to generate an empty tensor. 
![image](https://user-images.githubusercontent.com/8285571/125716872-8ae14d9d-b084-4fae-abe2-3146819ca721.png)
![image](https://user-images.githubusercontent.com/8285571/125716919-740f53ad-0b14-4c65-943c-09ebe0b64b54.png)
5. During PRZS, I also have the following debug information, where `current_share` and `next_share` are both empty. 
![image](https://user-images.githubusercontent.com/8285571/125717100-92594a27-0311-4010-8783-b44190454762.png)
![image](https://user-images.githubusercontent.com/8285571/125717113-489acc7d-9069-42da-8863-f2300d920583.png)
